packages/data-service/src/graphqlObservable.ts

```diff
@@ -0,0 +1,317 @@
+      return throwObservable("Unsupported use of fragments");
+    }
+
+    if (!definition.selectionSet) {
+      return Observable.of(parent);
+    }
+
+    return definition.selectionSet.selections.reduce((acc, sel) => {
+      if (
+        sel.kind === Kind.FRAGMENT_SPREAD ||
+        sel.kind === Kind.INLINE_FRAGMENT
+      ) {
+        return throwObservable("Unsupported use of fragments in selection set");
+      }
+
+      const result = resolve(sel, context, parent, type);
+      const fieldName = (sel.alias || sel.name).value;
+
+      return acc.combineLatest(result, objectAppendWithKey(fieldName));
+    }, Observable.of({}));
+  }
+
+  function resolveArrayResults(
+    definition: SchemaNode,
+    context: object,
+    parents: any[],
+    parentType: GraphQLType | null
+  ) {
+    return parents.reduce((acc, result) => {
+      const nextType = getResultType(parentType, definition, result);
+      const resultObserver = resolveResult(
+        definition,
+        context,
+        result,
+        nextType
+      );
+
+      return acc.combineLatest(resultObserver, listAppend);
+    }, Observable.of([]));
+  }
+
+  function getField(
+    parentType: GraphQLType | null,
+    definition: SchemaNode
+  ): GraphQLField<any, any> | null {
+    // Go one level deeper into the query
+    if (parentType instanceof GraphQLObjectType && isFieldNode(definition)) {
+      const parentFields = parentType.getFields();
+      const fieldName = definition.name.value;
+
+      if (parentFields[fieldName]) {
+        return parentFields[definition.name.value];
+      }
+    }
+
+    // These cases should ideally at some point be not existant,
+    // but due to our partial implementation this loop-hole is needed
+    return null;
+  }
+
+  function getResultType(
+    parentType: GraphQLType | null,
+    definition: SchemaNode,
+    instance: any
+  ): GraphQLType | null {
+    const translateOperation = {
+      query: "Query",
+      mutation: "Mutation"
+    };
+
+    // Operation is given (query or mutation), returns a type
+    if (isOperationDefinition(definition)) {
+      return types[translateOperation[definition.operation]];
+    }
+
+    // Get one level deeper in the query nesting
+    const field = getField(parentType, definition);
+    if (field !== null) {
+      const fieldType = getNamedType(field.type);
+
+      // Make this abstract type concrete if possible
+      if (
+        fieldType instanceof GraphQLInterfaceType &&
+        fieldType.resolveType instanceof Function
+      ) {
+        // We currenlty only allow resolveType to return a GraphQLObjectType
+        // and we pass in the wrong values as we don't need this feature currently
+        // @ts-ignore
+        return getNamedType(fieldType.resolveType(instance));
+      } else {
+        return fieldType;
+      }
+    }
+
+    return null;
+  }
+
+  return resolve(doc.definitions[0], context, null, null).map((data: T) => ({
```

Comment:
I'd like to see this expression moved to line 80 to help readability. Function hoisting allows it to appear before `function resolve`